transport/http2_client.go

```diff
@@ -904,15 +904,28 @@ func (t *http2Client) handleRSTStream(f *http2.RSTStreamFrame) {
 	s.write(recvMsg{err: io.EOF})
 }
 
-func (t *http2Client) handleSettings(f *http2.SettingsFrame) {
+func (t *http2Client) handleSettings(f *http2.SettingsFrame, isFirst bool) {
 	if f.IsAck() {
 		return
 	}
 	var ss []http2.Setting
+	isMaxConcurrentStreamsMissing := true
 	f.ForeachSetting(func(s http2.Setting) error {
+		if s.ID == http2.SettingMaxConcurrentStreams {
+			isMaxConcurrentStreamsMissing = false
+		}
 		ss = append(ss, s)
 		return nil
 	})
+	if isFirst && isMaxConcurrentStreamsMissing {
+		// This means server is imposing no limits on
+		// maximum number of concurrent streams initiated by client.
+		// So we must remove our self-imposed limit.
+		ss = append(ss, http2.Setting{
```

Comment:
This seems fishy to me.  But not so much _this_ as our settings handling in general.

It seems the way it works right now, we aren't _applying_ any settings until right as we send the ack.  This will result in a race if I'm understanding it correctly:

1. Server and client are in steady state with maximum streams = 5
2. Client receives a settings frame with a limit of 1.  It appends the ack to the control buffer in response.
3. Client initiates 5 new streams and pushes their headers onto the control buffer, which it thinks is fine because the settings are not applied yet.
4. Client applies settings and sends settings ack, promising not to initiate more than one stream.
5. Client sends headers for 5 new streams.

I think we need to make sure settings are applied _before_ queuing the ack message if the settings are more restrictive than they were before.  On the other hand, we should queue the ack message and then apply the settings if they are more permissive.  And...since it's possible for one setting to become more permissive while another becomes less permissive, we probably need to hold a lock, apply the settings, queue the ack, and then release the lock.